In [ ]:
!pip install -r '/kaggle/input/gptqmain/GPTQModel-main/requirements.txt'

In [ ]:
!cp -r /kaggle/input/klawq-gptq /kaggle/working/gptqvanilla2

In [4]:
import sys
import os

library_path = "/kaggle/working/gptqvanilla2" 
if library_path not in sys.path:
     sys.path.insert(0, library_path)
     print(f"Added '{library_path}' to sys.path")

gptq_file_path = os.path.join(library_path, 'gptqmodel', 'quantization', 'gptq.py') 
if os.path.exists(gptq_file_path):
    print(f"Found gptq.py at: {gptq_file_path}")
from gptqmodel import GPTQModel, QuantizeConfig 


Added '/kaggle/working/gptqvanilla2' to sys.path
Found gptq.py at: /kaggle/working/gptqvanilla2/gptqmodel/quantization/gptq.py

INFO  ENV: Auto setting PYTORCH_CUDA_ALLOC_CONF='expandable_segments:True' for memory saving.      
INFO  ENV: Auto setting CUDA_DEVICE_ORDER=PCI_BUS_ID for correctness.                              


In [ ]:

model_id   = "gpt2"                          
quant_path = "gpt2-text-calibrated" 
access_token = "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"

In [ ]:
from datasets import load_dataset
from itertools import islice

# 1) Open C4 in streaming mode (no full download)
c4_stream = load_dataset(
    "allenai/c4", 
    "en", 
    split="train", 
    streaming=True
)
num_calibration_samples = 1024
calibration_dataset_c4 = []
for sample in islice(c4_stream, num_calibration_samples):
    text = sample.get("text", "").strip()
    if not text:
        continue
    calibration_dataset_c4.append({"content": text})


print(f"Loaded {len(calibration_dataset_c4)} C4 examples for calibration.")
wt2 = load_dataset("wikitext", "wikitext-2-raw-v1", split="validation[:2000]")
eval_texts = [ex["text"] for ex in wt2 if ex["text"].strip()]

print(f"Using {len(eval_texts)} non-empty lines from Wikitext-2 for evaluation.")


In [ ]:
from transformers import AutoTokenizer
from gptqmodel import GPTQModel, QuantizeConfig
import os

torch.cuda.empty_cache()
gc.collect()

model_id    = "gpt2"
quant_path  = "/kaggle/working/gpt2-calibrated-8bit-g128"
quant_cfg   = QuantizeConfig(bits=8, group_size=128, beta=2.0, tau=0.7)
model       = GPTQModel.load(model_id, quant_cfg, trust_remote_code=True)
tokenizer = AutoTokenizer.from_pretrained(model_id, use_fast=True)
tokenizer.pad_token = tokenizer.eos_token
max_len = tokenizer.model_max_length

calib_tokenized = []
for txt in calibration_texts:
    enc = tokenizer(
        txt,
        truncation=True,
        padding="max_length",
        max_length=max_len,
    )
    calib_tokenized.append({
        "input_ids":      enc["input_ids"],
        "attention_mask": enc["attention_mask"],
    })

model.quantize(calib_tokenized, batch_size=8)
os.makedirs(os.path.dirname(quant_path), exist_ok=True)
model.save(quant_path)
print(f"Quantized GPT-2 saved to {quant_path}")


In [ ]:
import torch
import gc
def clear_gpu_cache():
    gc.collect()  
    torch.cuda.empty_cache()  
    torch.cuda.ipc_collect()  
    print("✅ GPU VRAM and cache cleared.")
clear_gpu_cache()

### Quantizd 8  bit, beta=2,temp - 0.7. AlienC4

In [6]:
import os
import math
import torch
from datasets import load_dataset
from transformers import AutoTokenizer
from gptqmodel import GPTQModel, QuantizeConfig
from itertools import islice

# 1) Stream a small “Alien-C4” subset
c4_stream = load_dataset("allenai/c4", "en", split="train", streaming=True)
eval_texts = [s["text"].strip() for s in islice(c4_stream, 1024) if s["text"].strip()]
print(f"Alien-C4 evaluation samples: {len(eval_texts)}")

# 2) Load the quantized GPT-2 model
quant_path = "/kaggle/working/gpt2-calibrated-8bit-g128"
quant_cfg  = QuantizeConfig(bits=8, group_size=128, beta=2.0, tau=0.7)
model = GPTQModel.from_pretrained(
    quant_path,
    trust_remote_code=True,
    device_map="auto",
    quantize_config=quant_cfg
)
model.eval()

# 3) Tokenize for evaluation
tokenizer           = AutoTokenizer.from_pretrained("gpt2", use_fast=True)
tokenizer.pad_token = tokenizer.eos_token
encodings = tokenizer(
    eval_texts,
    return_tensors="pt",
    padding="longest",
    truncation=True,
    max_length=tokenizer.model_max_length,
)
input_ids      = encodings.input_ids.to(model.device)
attention_mask = encodings.attention_mask.to(model.device)

# 4) Compute average loss & perplexity
batch_size = 8
losses = []
with torch.no_grad():
    for i in range(0, len(eval_texts), batch_size):
        b_ids  = input_ids[i : i+batch_size]
        b_mask = attention_mask[i : i+batch_size]
        out    = model(input_ids=b_ids, attention_mask=b_mask, labels=b_ids)
        losses.append(out.loss.item())

avg_loss   = sum(losses) / len(losses)
perplexity = math.exp(avg_loss)
print(f"\nAverage NLL loss: {avg_loss:.4f}")
print(f"Perplexity:       {perplexity:.2f}")

print("\n--- Sample Generations ---")
for prompt in eval_texts[:3]:
    enc = tokenizer(prompt, return_tensors="pt").to(model.device)
    gen = model.generate(**enc, max_new_tokens=50, do_sample=False)
    continuation = tokenizer.decode(
        gen[0, enc["input_ids"].size(1):],
        skip_special_tokens=True
    )
    print("Prompt:      ", prompt[:100] + "…")
    print("Continuation:", continuation.strip())
    print("-" * 40)

Resolving data files:   0%|          | 0/1024 [00:00<?, ?it/s]

Resolving data files:   0%|          | 0/1024 [00:00<?, ?it/s]

Alien-C4 evaluation samples: 1024
WARN  Model is already quantized, will use `from_quantized` to load quantized model.
If you want to quantize the model, please pass un_quantized model path or id, and use `from_pretrained` with `quantize_config`.
from_quantized: adapter: None
INFO  Loader: Auto dtype (native bfloat16): `torch.bfloat16`                                       
INFO  Estimated Quantization BPW (bits per weight): 8.31875 bpw, based on [bits: 8, group_size: 128]
INFO   Kernel: Auto-selection: adding candidate `TorchQuantLinear`                                 
INFO  Kernel: candidates -> `[TorchQuantLinear]`                                                   
INFO  Kernel: selected -> `TorchQuantLinear`.                                                      
INFO  Format: Converting `checkpoint_format` from `gptq` to internal `gptq_v2`.                    
INFO  Format: Conversion complete: 0.0025763511657714844s                                          
INFO   Kernel: Auto-se

### Quantized 8 bit, beta = 2, temp = 0.7, WikiText

In [5]:
import os, math, torch, pandas as pd
from datasets import load_dataset
from transformers import AutoTokenizer
from gptqmodel import GPTQModel, QuantizeConfig
from itertools import islice

# 1) Load your quantized GPT-2
quant_path = "/kaggle/working/gpt2-calibrated-8bit-g128"
quant_cfg  = QuantizeConfig(bits=8, group_size=128, beta=2.0, tau=0.7)
model = GPTQModel.from_pretrained(
    quant_path,
    trust_remote_code=True,
    device_map="auto",
    quantize_config=quant_cfg
)
model.eval()

# 2) Pull a small Wikitext-2 slice (128 non-empty lines)
wt2      = load_dataset("wikitext", "wikitext-2-raw-v1", split="validation")
texts    = [ex["text"].strip() for ex in wt2 if ex["text"].strip()][:128]
print(f"Using {len(texts)} Wikitext-2 lines for evaluation")

# 3) Tokenize with padding
tokenizer           = AutoTokenizer.from_pretrained("gpt2", use_fast=True)
tokenizer.pad_token = tokenizer.eos_token
encodings = tokenizer(
    texts,
    return_tensors="pt",
    padding="longest",
    truncation=True,
    max_length=tokenizer.model_max_length,
)
input_ids      = encodings.input_ids.to(model.device)
attention_mask = encodings.attention_mask.to(model.device)

# 4) Compute 4 metrics: NLL loss, perplexity, top-1 & top-5 token accuracy
batch_size = 8
losses, top1_accs, top5_accs = [], [], []

with torch.no_grad():
    for i in range(0, len(texts), batch_size):
        b_ids  = input_ids[i:i+batch_size]
        b_mask = attention_mask[i:i+batch_size]
        out    = model(input_ids=b_ids, attention_mask=b_mask, labels=b_ids)
        losses.append(out.loss.item())

        logits      = out.logits            # [B, L, V]
        sl_logits   = logits[:, :-1, :]     # shift for next-token pred
        sl_labels   = b_ids[:, 1:]
        sl_mask     = b_mask[:, 1:].bool()

        # top-1
        preds1      = sl_logits.argmax(dim=-1)
        correct1    = (preds1 == sl_labels) & sl_mask
        top1_accs.append(correct1.sum().item() / sl_mask.sum().item())

        # top-5
        topk        = 5
        topk_inds   = sl_logits.topk(topk, dim=-1).indices
        match5      = (topk_inds == sl_labels.unsqueeze(-1)) & sl_mask.unsqueeze(-1)
        top5_accs.append(match5.any(dim=-1).sum().item() / sl_mask.sum().item())

avg_loss   = sum(losses) / len(losses)
perplexity = math.exp(avg_loss)
acc1       = sum(top1_accs) / len(top1_accs)
acc5       = sum(top5_accs) / len(top5_accs)

# 5) Show metrics in a table
df = pd.DataFrame([
    {"Metric": "Avg NLL Loss",      "Value": avg_loss},
    {"Metric": "Perplexity",        "Value": perplexity},
    {"Metric": "Top-1 Token Acc",   "Value": acc1},
    {"Metric": "Top-5 Token Acc",   "Value": acc5},
])
print(df.to_markdown(index=False))

# 6) Print 3 sample continuations
print("\n--- Generations on Wikitext-2 ---")
for txt in texts[:3]:
    enc = tokenizer(txt, return_tensors="pt").to(model.device)
    gen = model.generate(**enc, max_new_tokens=50, do_sample=False)
    cont = tokenizer.decode(gen[0, enc["input_ids"].shape[1]:], skip_special_tokens=True)
    print("Prompt:      ", txt[:80] + "…")
    print("Continuation:", cont.strip())
    print("-" * 40)


WARN  Model is already quantized, will use `from_quantized` to load quantized model.
If you want to quantize the model, please pass un_quantized model path or id, and use `from_pretrained` with `quantize_config`.
from_quantized: adapter: None
INFO  Loader: Auto dtype (native bfloat16): `torch.bfloat16`                                       
INFO  Estimated Quantization BPW (bits per weight): 8.31875 bpw, based on [bits: 8, group_size: 128]
INFO   Kernel: Auto-selection: adding candidate `TorchQuantLinear`                                 
INFO  Kernel: candidates -> `[TorchQuantLinear]`                                                   
INFO  Kernel: selected -> `TorchQuantLinear`.                                                      
INFO  Format: Converting `checkpoint_format` from `gptq` to internal `gptq_v2`.                    
INFO  Format: Converting GPTQ v1 to v2                                                             
INFO  Format: Conversion complete: 0.004240274429321289s

In [2]:
import torch, math
import pandas as pd
from datasets import load_dataset
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig

# 1) Load GPT-2 with 8-bit quantization using bitsandbytes
quant_config = BitsAndBytesConfig(
    load_in_8bit=True,
    llm_int8_threshold=6.0,
    llm_int8_skip_modules=None
)
model = AutoModelForCausalLM.from_pretrained(
    "gpt2",
    quantization_config=quant_config,
    device_map="auto"
)
tokenizer = AutoTokenizer.from_pretrained("gpt2")
tokenizer.pad_token = tokenizer.eos_token

# 2) Load 128 Wikitext-2 non-empty lines
wt2 = load_dataset("wikitext", "wikitext-2-raw-v1", split="validation")
texts = [ex["text"].strip() for ex in wt2 if ex["text"].strip()][:128]
print(f"Using {len(texts)} Wikitext-2 lines for evaluation")

# 3) Tokenize
encodings = tokenizer(
    texts,
    return_tensors="pt",
    padding="longest",
    truncation=True,
    max_length=tokenizer.model_max_length,
)
input_ids = encodings.input_ids.to(model.device)
attention_mask = encodings.attention_mask.to(model.device)

# 4) Evaluate metrics
batch_size = 8
losses, top1_accs, top5_accs = [], [], []

model.eval()
with torch.no_grad():
    for i in range(0, len(texts), batch_size):
        b_ids = input_ids[i:i+batch_size]
        b_mask = attention_mask[i:i+batch_size]
        out = model(input_ids=b_ids, attention_mask=b_mask, labels=b_ids)
        losses.append(out.loss.item())

        logits = out.logits[:, :-1, :]
        sl_labels = b_ids[:, 1:]
        sl_mask = b_mask[:, 1:].bool()

        preds1 = logits.argmax(dim=-1)
        correct1 = (preds1 == sl_labels) & sl_mask
        top1_accs.append(correct1.sum().item() / sl_mask.sum().item())

        top5 = logits.topk(5, dim=-1).indices
        correct5 = (top5 == sl_labels.unsqueeze(-1)) & sl_mask.unsqueeze(-1)
        top5_accs.append(correct5.any(dim=-1).sum().item() / sl_mask.sum().item())

# 5) Report metrics
avg_loss = sum(losses) / len(losses)
perplexity = math.exp(avg_loss)
acc1 = sum(top1_accs) / len(top1_accs)
acc5 = sum(top5_accs) / len(top5_accs)

df = pd.DataFrame([
    {"Metric": "Avg NLL Loss", "Value": avg_loss},
    {"Metric": "Perplexity", "Value": perplexity},
    {"Metric": "Top-1 Token Acc", "Value": acc1},
    {"Metric": "Top-5 Token Acc", "Value": acc5},
])
print(df.to_markdown(index=False))

# 6) Generate sample continuations
print("\n--- Generations on Wikitext-2 ---")
for txt in texts[:3]:
    enc = tokenizer(txt, return_tensors="pt").to(model.device)
    gen = model.generate(**enc, max_new_tokens=50, do_sample=False)
    cont = tokenizer.decode(gen[0, enc["input_ids"].shape[1]:], skip_special_tokens=True)
    print("Prompt:      ", txt[:80] + "…")
    print("Continuation:", cont.strip())
    print("-" * 40)


Using 128 Wikitext-2 lines for evaluation


Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


| Metric          |       Value |
|:----------------|------------:|
| Avg NLL Loss    |    8.50262  |
| Perplexity      | 4927.66     |
| Top-1 Token Acc |    0.355017 |
| Top-5 Token Acc |    0.563651 |

--- Generations on Wikitext-2 ---


Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


Prompt:       = Homarus gammarus =…
Continuation: Homarus gammarus = Homarus gammarus = Homarus gammarus = Homarus gammarus = Homarus gammarus = Homarus gammarus = Homarus gammarus = Homarus gammarus = Homarus
----------------------------------------


Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


Prompt:       Homarus gammarus , known as the European lobster or common lobster , is a specie…
Continuation: The species is found in the Mediterranean Sea and parts of the Black Sea . It is closely related to the American lobster , H. americanus . It may grow to a length of 60 cm ( 24 in ) and a mass of 6 kilograms
----------------------------------------
Prompt:       = = Description = =…
Continuation: The name of the file to be used.

The name of the file to be used. The name of the file to be used. The name of the file to be used. The name of the file to be used. The
----------------------------------------


### Analysis ###

## Evaluation Comparison: Custom GPTQ vs. HF 8-bit GPT-2

Below is a comparison between our custom GPTQ-quantized GPT-2 model (β = 2.0, τ = 0.7) and the default 8-bit GPT-2 model from the Hugging Face library, evaluated on 128 lines from the Wikitext-2 validation set.

| Metric             | GPTQ (β=2.0, τ=0.7) | HF 8-bit GPT-2 |
|--------------------|--------------------:|----------------:|
| Avg NLL Loss       |              8.4160 |          8.5026 |
| Perplexity         |           4518.89   |        4927.66  |
| Top-1 Token Acc.   |             35.18 % |         35.50 % |
| Top-5 Token Acc.   |             56.16 % |         56.36 % |

Our KL-Aware GPTQ model demonstrates better loss and perplexity, showing stronger probabilistic modeling of the text. While the Hugging Face model performs slightly better in top-1 and top-5 token accuracy, the difference is minimal. This suggests that our quantization settings preserve model confidence better, with only marginal trade-offs in exact token prediction. I believe this will require a thorough Beta, temperature values testing to give some final analysi
